File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        def twoSum(ans, curr, nums, left):
            counter = dict()
            i = left
            while i < len(nums):
                complement = -curr - nums[i]
                if complement in counter.keys():
                    ans.append([curr, nums[i], complement])
                    while i + 1 < len(nums) and nums[i + 1] == nums[i]:
                        i += 1
                counter[nums[i]] = complement
                i += 1
                    
        if len(nums) < 3:
            return []
        ans = []
        nums.sort()
        for i in range(len(nums) - 2):
            if nums[i] > 0:
                break
            if i == 0 or nums[i - 1] != nums[i]:
                curr = nums[i]
                left = i + 1
                twoSum(ans, curr, nums, left)
        return ans
        

File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        if (nums.length < 3) return ans;
        Arrays.sort(nums);
        int n = nums.length;
        for (int i = 0; i < n - 2; i++) {
            if (i == 0 || nums[i] != nums[i - 1]) {
                int complement = -1 * nums[i];
                int left = i + 1;
                int right = n - 1;
                while (left < right) {
                    if (nums[left] + nums[right] == complement) {
                        List<Integer> temp = new ArrayList<>();
                        temp.add(nums[i]);
                        temp.add(nums[left]);
                        temp.add(nums[right]);
                        ans.add(temp);
                        left++;
                        while (left < right && nums[left] == nums[left - 1]) {
                            left++;
                        }
                    } else if (nums[left] + nums[right] > complement) {
                        right--;
                    } else {
                        left++;
                    }
                }
            }
        }
        return ans;
    }
}